### PR TITLE
Rollouts package discovery

### PR DESCRIPTION
**Pull request by ChristopherFry**
This pull request adds package discovery to rollouts.

### CODE REVIEW

File: rollouts/pkg/packagediscovery/packagediscovery.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PackageDiscovery struct {
+	config    gitopsv1alpha1.PackagesConfig
+	client    client.Client
+	namespace string
+}
+
+type DiscoveredPackage struct {
```
Comment (droot) [inline rollouts/pkg/packagediscovery/packagediscovery.go:22]: Note to self: I am realizing we will need some consistent way of discovering `package id + version`.

File: rollouts/pkg/packagediscovery/packagediscovery.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func (d *PackageDiscovery) GetPackages(ctx context.Context) ([]DiscoveredPackage, error) {
+	gitRepoSelector := d.config.Git.GitRepoSelector
+	gitClient, err := d.getGitHubClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create git client: %w", err)
+	}
+
+	tree, _, err := gitClient.Git.GetTree(ctx, gitRepoSelector.Org, gitRepoSelector.Name, gitRepoSelector.Revision, true)
```
Comment (droot) [inline rollouts/pkg/packagediscovery/packagediscovery.go:44]: sweet!

Comment (droot) [reply rollouts/pkg/packagediscovery/packagediscovery.go:44]: Does it give content hash for given revision as well ?

File: rollouts/pkg/packagediscovery/packagediscovery.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+		httpClient = oauth2.NewClient(ctx, ts)
+	}
+
+	gitClient := github.NewClient(httpClient)
+
+	return gitClient, nil
+}
+
+func discoverPackagePaths(pattern string, paths []string) []string {
```
Comment (droot) [inline rollouts/pkg/packagediscovery/packagediscovery.go:94]: May be call this method `filterPaths` or something to make it obvious.

File: rollouts/pkg/packagediscovery/packagediscovery.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	return discoveredPackages, nil
+}
+
+func (d *PackageDiscovery) getGitHubClient(ctx context.Context) (*github.Client, error) {
+	gitRepoSelector := d.config.Git.GitRepoSelector
+
+	httpClient := &http.Client{}
+
+	if secretName := gitRepoSelector.SecretRef.Name; secretName != "" {
```
Comment (droot) [inline rollouts/pkg/packagediscovery/packagediscovery.go:73]: Thinking from API perspective, I think for `git` we can decouple the connectionInfo (auth in this case) from the different types of selectors. And we can do that as we evolve the code.

**Review by droot — APPROVED** [review]
Looks good to me. I have minor comments that we can address in the follow ups or as we evolve the code.

File: rollouts/pkg/packagediscovery/packagediscovery.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
+package packagediscovery
```
[the remark this answers is not in the document]
Comment (ChristopherFry) [reply rollouts/pkg/packagediscovery/packagediscovery.go:15]: Done.

File: rollouts/pkg/packagediscovery/packagediscovery.go (hunk 3 above)
Comment (ChristopherFry) [reply rollouts/pkg/packagediscovery/packagediscovery.go]: Done.

File: rollouts/api/v1alpha1/rollout_types.go [hunk 6]
```diff
@@ -30,20 +30,47 @@ type RolloutSpec struct {
...
+// GitSource defines the packages source in Git.
+type GitSource struct {
+	GitRepoSelector GitSelector `json:"selector"`
+}
+
+// GitSelector defines the selector to apply to Git.
+type GitSelector struct {
+	Org          string          `json:"org"`
+	Name         string          `json:"name"`
+	PackagesPath string          `json:"packagesPath"`
```
[the remark this answers is not in the document]
Comment (ChristopherFry) [reply rollouts/api/v1alpha1/rollout_types.go]: Correct. Also just realized we're handling directories here, not paths, so updated the name to 'Directory'.

File: rollouts/api/v1alpha1/rollout_types.go [hunk 7]
```diff
@@ -30,20 +30,47 @@ type RolloutSpec struct {
...
+// PackagesConfig defines the packages the Rollout should deploy.
+type PackagesConfig struct {
+	SourceType PackageSourceType `json:"sourceType"`
+
+	Git GitSource `json:"git"`
+}
+
+// GitSource defines the packages source in Git.
+type GitSource struct {
+	GitRepoSelector GitSelector `json:"selector"`
```
[the remark this answers is not in the document]
Comment (ChristopherFry) [reply rollouts/api/v1alpha1/rollout_types.go:56]: Both. Today this only supports directory selectors today however it'll be able to support repo-selectors in the future as well. Essentially the controller will inspect the fields of this GitSelector, and will use selectors for any fields that contain a wildcard.

**Pull request merged**